internal/controller/nodetemplate_controller.go

```diff
@@ -0,0 +1,390 @@
+	}
+	node.Labels["node-role.kubernetes.io"+"/"+nodeTemplate.Name] = "true"
+	for k, v := range nodeTemplate.Spec.Template.Metadata.Labels {
+		node.Labels[k] = v
+	}
+	if nodeTemplate.Spec.Template.Metadata.Annotations != nil && node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+	for k, v := range nodeTemplate.Spec.Template.Metadata.Annotations {
+		node.Annotations[k] = v
+	}
+}
+
+func selectNodes(spareNodes map[string]corev1.Node, zoneCount map[string]int) corev1.Node {
+	maxCountPerZone := 100
+	maxScore := -1
+	maxScoreNode := ""
+	for k, n := range spareNodes {
+		zoneCountScore := 0
+		if zone, ok := n.Labels[ZoneLabelKey]; ok {
+			zoneCountScore = maxCountPerZone - zoneCount[zone]
+		}
+		score := zoneCountScore * 10
+		if score > maxScore {
+			maxScore = score
+			maxScoreNode = k
+		}
+	}
+	return spareNodes[maxScoreNode]
+}
+
+func (r *NodeTemplateReconciler) updateStatus(ctx context.Context, nodeTemplate *nyallocatorv1.NodeTemplate) error {
+	nodes := corev1.NodeList{}
+	err := r.Client.List(ctx, &nodes, client.MatchingLabels{NodeTemplateReferenceLabelKey: nodeTemplate.Name})
+	if err != nil {
+		return err
+	}
+	nodeTemplate.Status.ReconcileInfo.Generation = nodeTemplate.Generation
+	nodeTemplate.Status.CurrentNodes = len(nodes.Items)
+	nodeTemplate.Status.Sufficient = nodeTemplate.Status.CurrentNodes >= nodeTemplate.Spec.Nodes
+	r.updateMetrics(nodeTemplate)
+	err = r.Client.Status().Update(ctx, nodeTemplate)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *NodeTemplateReconciler) updateMetrics(nodeTemplate *nyallocatorv1.NodeTemplate) {
+	if nodeTemplate.Status.Sufficient {
+		SufficientNodesVec.WithLabelValues(nodeTemplate.Name).Set(1)
+	} else {
+		SufficientNodesVec.WithLabelValues(nodeTemplate.Name).Set(0)
+	}
+	CurrentNodesVec.WithLabelValues(nodeTemplate.Name).Set(float64(nodeTemplate.Status.CurrentNodes))
+	ExpectedNodesVec.WithLabelValues(nodeTemplate.Name).Set(float64(nodeTemplate.Spec.Nodes))
+}
+
+func (r *NodeTemplateReconciler) removeMetrics(nodeTemplate *nyallocatorv1.NodeTemplate) {
+	SufficientNodesVec.DeleteLabelValues(nodeTemplate.Name)
+	CurrentNodesVec.DeleteLabelValues(nodeTemplate.Name)
+	ExpectedNodesVec.DeleteLabelValues(nodeTemplate.Name)
+}
+
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;update;patch
+
+func (r *NodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	node := corev1.Node{}
+	err := r.Client.Get(ctx, client.ObjectKey{Name: req.Name}, &node)
+	if err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == SpareTaintKey {
+			if _, ok := node.Labels[SpareRoleLabelKey]; !ok {
+				// if spare node but does not have spare role label, add it
+				node.Labels[SpareRoleLabelKey] = "true"
+				err := r.Client.Update(ctx, &node)
+				if err != nil {
+					return ctrl.Result{}, err
```

Comment:
We may ignore NotFound here as well.
```suggestion
					return ctrl.Result{}, client.IgnoreNotFound(err)
```